### ISSUE TITLE
BufferUnderflowException in case of using java 8 with lambdas

### ISSUE THREAD

**Issue by coiouhkc** [payara/Payara#959]
As described in 
http://stackoverflow.com/questions/28301584/javaee-glassfish-bufferunderflowexception - in case an EJB Module is using Lambdas, ConstantPoolInfo first fails to recognize 18 (INVOKEDYNAMIC, see http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.4) and then goes all wrong.
We have verified the fix proposed by Koen Serneels in the above SO thread.
We're  currently using Payara 4.1.1.162.
Is the fix for the above issue already included in some later branch/release, or are there an plans to include it?
In case you would like to include it by didn't yet have time for it - please let me know. I'd be happy to create a pull request.

**Comment by OndrejM** [comment]
There is a bug open (unresolved) for GlassFish already: https://java.net/jira/browse/GLASSFISH-21510
From the comments there, it is not exactly clear how to fix the bug. 
We've had another issue #689 related to problems with new constructs in Java 8 , but it was fixed in 4.1.1.162 release already, so it does not seem to fix this issue.

Since Payara 163 is very close to its release, it is highly unlikely that a fix for this issue will be included. 
However, we understand how serious this issue can be and will treat it ASAP after the next release.

An EJB-JAR file that can be used **to reproduce the problem** would be very helpful.

We would also gladly **welcome a pull request** with a fix. You would need to sign a [contributor agreement](https://github.com/payara/Payara/blob/master/Contributing.md) before we can accept it though.

**Comment by coiouhkc** [comment]
Thanks for prompt reply.
The problem seems to occur (in our case) during test execution with embedded Payara whilst havinf deployed EJB artifacts containing lambdas - i'll try to create a test project to reproduce the problem.

**Comment by coiouhkc** [comment]
Here is the sample project which (although not throwing the BufferUnderflowException) depicts the issue: https://github.com/coiouhkc/payara-issue-959.

**Comment by coiouhkc** [comment]
Here a full stacktrace from our project with BufferUnderflowException:
java.nio.BufferUnderflowException
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:151)
	at com.sun.enterprise.deployment.annotation.introspection.ConstantPoolInfo.containsAnnotation(ConstantPoolInfo.java:86)
	at com.sun.enterprise.deployment.annotation.introspection.ClassFile.containsAnnotation(ClassFile.java:133)
	at com.sun.enterprise.deployment.annotation.introspection.ClassFile.containsAnnotation(ClassFile.java:87)
	at com.sun.enterprise.deployment.util.AnnotationDetector.containsAnnotation(AnnotationDetector.java:157)
	at com.sun.enterprise.deployment.util.AnnotationDetector.containsAnnotation(AnnotationDetector.java:146)
	at com.sun.enterprise.deployment.util.AnnotationDetector.hasAnnotationInArchive(AnnotationDetector.java:137)
	at com.sun.enterprise.deployment.archivist.ApplicationArchivist.getApplicationFromIntrospection(ApplicationArchivist.java:393)
	at com.sun.enterprise.deployment.archivist.ApplicationArchivist.createApplication(ApplicationArchivist.java:264)
	at org.glassfish.javaee.full.deployment.EarHandler.getApplicationHolder(EarHandler.java:578)
	at org.glassfish.javaee.full.deployment.EarHandler.initCompositeMetaData(EarHandler.java:555)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.getSniffers(ApplicationLifecycle.java:655)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:358)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:219)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:491)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
	at com.sun.enterprise.admin.cli.embeddable.DeployerImpl.deploy(DeployerImpl.java:133)
	at org.glassfish.ejb.embedded.EJBContainerImpl.deploy(EJBContainerImpl.java:141)
	at org.glassfish.ejb.embedded.EJBContainerProviderImpl.createEJBContainer(EJBContainerProviderImpl.java:134)
	at javax.ejb.embeddable.EJBContainer.createEJBContainer(EJBContainer.java:127)

**Comment by coiouhkc** [comment]
Added pull request: https://github.com/payara/Payara/pull/965.

**Comment by smillidge** [comment]
I can only reproduce this with your test case specifically if I create an EJB jar with your bean in I can deploy to Payara Server fine.

**Comment by smillidge** [comment]
strange if you remove the line props.put(EJBContainer.APP_NAME, "ejb-app"); there is no failure

**Comment by coiouhkc** [comment]
Interesting, how is it related to 'ejb-app'?
java:global[/<app-name>]/<module-name>/<bean-name> should be a totally valid JNDI lookup for EJB 3.1 with <app-name> = "ejb-app"..

**Comment by coiouhkc** [comment]
Thanks!
